File: app/src/main/java/com/doubean/ford/ui/groups/groupsHome/GroupsHomeFragment.java
package com.doubean.ford.ui.groups.groupsHome;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;

import com.doubean.ford.R;
import com.doubean.ford.adapters.GroupFollowAdapter;
import com.doubean.ford.adapters.RecommendedGroupAdapter;
import com.doubean.ford.databinding.FragmentGroupsBinding;
import com.doubean.ford.util.InjectorUtils;

import java.util.ArrayList;

public class GroupsHomeFragment extends Fragment {

    private GroupsHomeViewModel groupsHomeViewModel;
    private GroupsHomeViewModelFactory factory;
    private FragmentGroupsBinding binding;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        factory = InjectorUtils.provideGroupsViewModelFactory(requireContext());
        groupsHomeViewModel =
                new ViewModelProvider(this, factory).get(GroupsHomeViewModel.class);
        binding = FragmentGroupsBinding.inflate(inflater, container, false);
        binding.setViewModel(groupsHomeViewModel);
        binding.setLifecycleOwner(getViewLifecycleOwner());

        GroupFollowAdapter adapter = new GroupFollowAdapter();
        binding.followList.setAdapter(adapter);
        groupsHomeViewModel.getFollowList().observe(getViewLifecycleOwner(), list -> adapter.submitList(list == null ? null : new ArrayList<>(list)));

        RecommendedGroupAdapter recommendedGroupAdapter = new RecommendedGroupAdapter(getContext());
        binding.groupsOfTheDayList.setAdapter(recommendedGroupAdapter);
        groupsHomeViewModel.getGroupsOfTheDay().observe(getViewLifecycleOwner(), groups -> recommendedGroupAdapter.submitList(groups == null || groups.data == null ? null : new ArrayList<>(groups.data)));
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.toolbar.inflateMenu(R.menu.groups_menu);
        binding.toolbar.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.action_search: {
                    NavDirections action = GroupsHomeFragmentDirections.actionNavigationGroupsToNavigationGroupSearch();
                    Navigation.findNavController(view).navigate(action);
                    return true;
                }
                case R.id.action_settings:
                    Navigation.findNavController(view).navigate(R.id.navigation_settings_fragment);
                    return true;
                default:
                    return false;
            }
        });
    }

}

File: app/src/main/java/com/doubean/ford/ui/groups/groupsHome/GroupsHomeFragment.kt
package com.doubean.ford.ui.groups.groupsHome

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.doubean.ford.R
import com.doubean.ford.adapters.RecommendedGroupAdapter
import com.doubean.ford.databinding.FragmentGroupsBinding
import com.doubean.ford.util.InjectorUtils

class GroupsHomeFragment : Fragment() {
    private val groupsHomeViewModel: GroupsHomeViewModel by viewModels {
        InjectorUtils.provideGroupsViewModelFactory(requireContext())
    }
    private lateinit var binding: FragmentGroupsBinding
    private lateinit var followAdapter: GroupFollowAdapter
    private lateinit var recommendedGroupAdapter: RecommendedGroupAdapter
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?,
    ): View {
        binding = FragmentGroupsBinding.inflate(inflater, container, false).apply {
            viewModel = groupsHomeViewModel
            lifecycleOwner = viewLifecycleOwner
        }
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.toolbar.inflateMenu(R.menu.groups_menu)
        binding.toolbar.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_search -> {
                    val action =
                        GroupsHomeFragmentDirections.actionGroupsToGroupSearch()
                    findNavController().navigate(action)
                    true
                }
                R.id.action_settings -> {
                    findNavController().navigate(R.id.navigation_settings)
                    true
                }
                else -> false
            }
        }

        followAdapter = GroupFollowAdapter()
        binding.followList.adapter = followAdapter

        recommendedGroupAdapter = RecommendedGroupAdapter(requireContext())
        binding.groupsOfTheDayList.adapter = recommendedGroupAdapter

        groupsHomeViewModel.follows
            .observe(viewLifecycleOwner) { followAdapter.submitList(it ?: emptyList()) }
        groupsHomeViewModel.groupsOfTheDay.observe(viewLifecycleOwner) { groups ->
            recommendedGroupAdapter.submitList(groups.data ?: emptyList())
        }
    }

}